README.pod
=pod

=encoding UTF-8

=head1 NAME

MarpaX::Languages::ECMAScript::AST - Translate a ECMAScript source to an AST

=head1 VERSION

version 0.019

=head1 SYNOPSIS

    use strict;
    use warnings FATAL => 'all';
    use MarpaX::Languages::ECMAScript::AST;
    use Log::Log4perl qw/:easy/;
    use Log::Any::Adapter;
    use Log::Any qw/$log/;
    #
    # Init log
    #
    our $defaultLog4perlConf = '
    log4perl.rootLogger              = WARN, Screen
    log4perl.appender.Screen         = Log::Log4perl::Appender::Screen
    log4perl.appender.Screen.stderr  = 0
    log4perl.appender.Screen.layout  = PatternLayout
    log4perl.appender.Screen.layout.ConversionPattern = %d %-5p %6P %m{chomp}%n
    ';
    Log::Log4perl::init(\$defaultLog4perlConf);
    Log::Any::Adapter->set('Log4perl');
    #
    # Parse ECMAScript
    #
    my $ecmaSourceCode = 'var i = 0;';
    my $ecmaAstObject = MarpaX::Languages::ECMAScript::AST->new();
    $log->infof('%s', $ecmaAstObject->parse($ecmaSourceCode));

=head1 DESCRIPTION

This module translates ECMAScript source into an AST tree. If you want to enable logging, be aware that this module is using Log::Any.

=head1 SUBROUTINES/METHODS

=head2 new($class, %options)

Instantiate a new object. Takes as parameter an optional hash of options that can be:

=over

=item grammarName

Name of a grammar. Default is 'ECMAScript-262-5'.

=item cache

Produced AST can be cached: very often the same ECMAScript is used again and again, so there is no need to always compute it at each call. The cache key is the buffer MD4 checksum, eventual collisions being handled. The cache location is the my_dist_data directory provided by File::HomeDir package. Default is a false value.

=item ANYTHINGELSE

Any other option that will be specific to the ECMAScript specific implementation. In practice these are:

=over

=item Template

Reference to a hash for template options. C.f. per-ECMAScript documentation.

=item StringNumericLiteral

Reference to a hash for StringNumericLiteral grammar options. C.f. per-ECMAScript documentation.

=item Pattern

Reference to a hash for Pattern grammar options. C.f. per-ECMAScript documentation.

=back

=back

=head2 describe($self)

Get a description of the G1 grammar. Returns a reference to hash, that has this structure: key => value, where

=over

=item key

'G1' or 'G0', with value being a reference to all rules, LHS being the first.

=back

=head2 grammarAlias($self)

Returns the grammar alias, i.e. the one that is used within this distribution, corresponding to a true directory on the filesystem.

=head2 templatePath($self)

Returns the templates absolute path.

=head2 parse($self, $source)

Get an AST from the ECMAScript source, pointed by $source. This method will call all the intermediary steps (lexical, transformation, evaluation) necessary to produce the AST. An AST is a reference to nested arrays, containing either:

=over

=item valuesAndRuleId

An entry describing a G1 rule (i.e. a non-terminal) is a reference to an array. The array contains the values of the right-hand side of the rule, followed by the rule Id.

=item startAndLengthAndValue

An entry describing a lexeme (i.e. a terminal) is a reference to an array. The array contains the start location in the stream, length and lexeme as perl's utf8 string.

=back

=head2 template($self)

Return the generic template for this grammar. This template is doing nothing else but reproduce an ECMAScript source that, if parsed, would have an AST similar to the original source.

=head2 stringNumericLiteral($self)

Return the generic stringNumericLiteral for this grammar.

=head2 pattern($self)

Return the generic Pattern for this grammar.

=head2 JSON($self)

Return the generic JSON for this grammar.

=head2 URI($self)

Return the generic URI for this grammar.

=head1 SEE ALSO

L<Log::Any>

L<Marpa::R2>

L<Digest::MD4>

L<CHI::Driver::File>

L<MarpaX::Languages::ECMAScript::AST::Grammar::ECMAScript_262_5>

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://rt.cpan.org/Public/Dist/Display.html?Name=MarpaX-Languages-ECMAScript-AST>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/jddurand/marpax-languages-ecmascript-ast>

  git clone git://github.com/jddurand/marpax-languages-ecmascript-ast.git

=head1 AUTHOR

Jean-Damien Durand <jeandamiendurand@free.fr>

=head1 CONTRIBUTOR

=for stopwords jddurand

jddurand <jeandamiendurand@free.fr>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Jean-Damien Durand.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Changelog for MarpaX-Languages-ECMAScript-AST

0.019 2014-10-25T22:15:15
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - ECMAScript_262_5/Program.pm: github issue #5
 - t/asi.t: say more when there is a failure

0.018 2014-07-23T20:09:04Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.018
 - t/github_issue_0004.t: More than one parse tree value #4
 - t/github_issue_0003.t: fix comment
 - github issue #3
 - ECMAScript_262_5 Generic Template for version 0.018
 - Grammar/Base.pm: too_many_earley_items as a class method
 - Grammar/Base.pm: too_many_earley_items => 0
 - ECMASCript/Program.pm: fix ExpressionStatement ambiguity with Block or
   FunctionDeclaration
 - Grammar/Base.pm: support of eventual traverser
 - ECMAScript_262_5 Generic Template for version 0.018
 - AST/Impl.pm: fix recce method

0.017 2014-07-03T21:41:45Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.017
 - ECMAScript_262_5 Generic Template for version 0.017
 - Program/Semantics.pm: optimization
 - Merge branch 'master' of
   https://github.com/jddurand/MarpaX-Languages-ECMAScript-AST
 - Program/Semantics.pm: add lhs and rhs information in value -
   StringLiteral: documentation and behaviour mismatch #2
 - ECMAScript_262_5/Program.pm: few optimizations
 - ECMAScript_262_5 Generic Template for version 0.017
 - t/github_issue_0001.t: Removed logging

0.016 2014-07-01T22:45:04Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.016
 - Code that throws a "subexpression recursion limit exception". (#1)
 - number dev+

0.015 2014-02-22T08:34:45Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.015
 - Grammar/Base.pm: Fixed show of last expression in case of error
 - Revert "Use slr->ambiguity_metric to detect ambiguous parse" This
   reverts commit 510867cf17ba5b65515f76af8ed34a33f7445659.
 - dev+
 - dev+
 - dev+
 - dev+
 - Merge branch 'master' of
   https://github.com/jddurand/MarpaX-Languages-ECMAScript-AST
 - AST/Impl.pm: use Marpa::R2 2.080000
 - Use slr->ambiguity_metric to detect ambiguous parse

0.014 2014-02-16T21:10:23Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.014
 - Pattern/Semantics.pm: revisited way to condtionnaly use perl feature

0.013 2014-02-16T15:21:39Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.013
 - use Marpa::R2 2.079015 rewritten as use Marpa::R2 2.079_015
 - t/pattern.t: full coverage of Pattern/Semantics.pm
 - Pattern/Semantics.pm: _HexEscapeSequence fix
 - t/pattern.c: escape parens test
 - lexeme default = forgiving => 1 && DecimalDigits
 - lexeme default = forgiving => 1
 - t/pattern.t: Caret test not at the beginning of string
 - Pattern/Semantics.pm: Caret assertino bug
 - Pattern/Semantics.pm: bug in _charsetUnion
 - t/pattern.t: more tests
 - Pattern/Semantics.pm: one more parameter for regexp, a routine to
   eventually do its own upper case
 - t/pattern.t: Assertion \b and \B tests conted
 - t/pattern.t: Assertion \\b and \\B tests
 - Pattern/Semantics.pm: bug fix in _isWordChar
 - script/cover.sh
 - _AtomEscape_CharacterClassEscape fix

0.012 2014-02-14T21:21:11Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.012
 - use File::HomeDir 0.93; because of my_dist_data

0.011 2014-02-09T23:30:10Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.011
 - Impl/Singleton.pm: Removed some debug statement
 - t/stringNumericLiteralBigFloat.t: on some old Math::BigInt, 0x is not
   optional in from_hex

0.010 2014-02-09T01:01:49Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.010
 - Float wooes
 - StringNumericLiteral/NativeNumberSemantics.pm: do {my $n = 2; $n *= $n
   while $n < $n*$n; $n} only when absolutely necessary

0.009 2014-02-08T18:16:41Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.009
 - t/stringNumericLiteral*.t: Infinity tests
 - StringNumericLiteral/NativeNumberSemantics.pm: Try to deal with
   platforms without pos_zero or pos_infinity
 - ECMAScript_262_5/Program.pm: Tiny optimization

0.008 2014-02-07T19:36:39Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.008
 - ECMAScript_262_5 Generic Template for version 0.008
 - Replaced __PACKAGE__ by $class
 - Fix tests
 - Problem with file naming (2)
 - Pb with file naming
 - Fix Semantics package runtime settings
 - Added missing ABSTRACT and VERSION
 - URI prototype
 - URI dev+
 - JSON prototype
 - dist.ini: double quotes for portability
 - Singleton centralization
 - AST/Impl.pm: Typo in POD
 - Pattern/Semantics.pm: Fix POD error

0.007 2014-02-05T23:04:05Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.007
 - Refactorisation
 - Lexical/NumericLiteral.pm: Load grammar once
 - Program/Actions.pm renamed to Program/DefaultSemanticsPackage.pm
 - StringNumericLiteral: fix bareword warnings
 - t/pattern.t: better indentation
 - t/pattern.t: move test of value upper
 - t/pattern.t: two more test with alternation
 - Pattern: Fix lexeme issue (longest match win)

0.006 2014-02-03T22:18:19Z
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Release of version 0.006
 - use
   MarpaX::Languages::ECMAScript::AST::Grammar::ECMAScript_262_5::Character
   Classes
 - t/pattern.t: Removed logging
 - Pattern: t+ and deviation from the spec for Assertions returning a
   matcher
 - Pattern: t+
 - Revert "ECMAScript_262_5/Pattern.pm :  missing track of lparens for
   lookahead disjunctions" This reverts commit
   71d04f099f953fde146afd4ec3d5b4b59279fa85.
 - Revert "Pattern/DefaultSemanticsPackage.pm: support of lparens in
   lookahead disjunctions" This reverts commit
   72cfe8ac1f5a430bbc0e6e3d9820786ba53e3fbe.
 - Pattern/DefaultSemanticsPackage.pm: support of lparens in lookahead
   disjunctions
 - ECMAScript_262_5/Pattern.pm :  missing track of lparens for lookahead
   disjunctions
 - Pattern: positive lookahead test
 - Pattern: backtracking fix

lib/MarpaX/Languages/ECMAScript/AST.pm
use strict;
use warnings FATAL => 'all';

package MarpaX::Languages::ECMAScript::AST;

# ABSTRACT: Translate a ECMAScript source to an AST

use MarpaX::Languages::ECMAScript::AST::Grammar qw//;
use Digest::MD4 qw/md4_hex/;
use CHI;
use File::HomeDir 0.93;
use version 0.77;
use Log::Any qw/$log/;
use Module::Util qw/find_installed/;
use File::Basename qw/dirname/;
use File::Spec;

our $distname = __PACKAGE__;
$distname =~ s/::/-/g;

our $CACHE = CHI->new(driver => 'File',
                      root_dir => File::HomeDir->my_dist_data($distname, { create => 1 } ),
                      label => __PACKAGE__,
                      namespace => 'cache',
		      max_key_length => 32);

# VERSION
our $CURRENTVERSION;
{
  #
  # Because $VERSION is generated by dzil, not available in dev. tree
  #
  no strict 'vars';
  $CURRENTVERSION = $VERSION;
}

=head1 DESCRIPTION

This module translates ECMAScript source into an AST tree. If you want to enable logging, be aware that this module is using Log::Any.

=head1 SYNOPSIS

    use strict;
    use warnings FATAL => 'all';
    use MarpaX::Languages::ECMAScript::AST;
    use Log::Log4perl qw/:easy/;
    use Log::Any::Adapter;
    use Log::Any qw/$log/;
    #
    # Init log
    #
    our $defaultLog4perlConf = '
    log4perl.rootLogger              = WARN, Screen
    log4perl.appender.Screen         = Log::Log4perl::Appender::Screen
    log4perl.appender.Screen.stderr  = 0
    log4perl.appender.Screen.layout  = PatternLayout
    log4perl.appender.Screen.layout.ConversionPattern = %d %-5p %6P %m{chomp}%n
    ';
    Log::Log4perl::init(\$defaultLog4perlConf);
    Log::Any::Adapter->set('Log4perl');
    #
    # Parse ECMAScript
    #
    my $ecmaSourceCode = 'var i = 0;';
    my $ecmaAstObject = MarpaX::Languages::ECMAScript::AST->new();
    $log->infof('%s', $ecmaAstObject->parse($ecmaSourceCode));

=head1 SUBROUTINES/METHODS

=head2 new($class, %options)

Instantiate a new object. Takes as parameter an optional hash of options that can be:

=over

=item grammarName

Name of a grammar. Default is 'ECMAScript-262-5'.

=item cache

Produced AST can be cached: very often the same ECMAScript is used again and again, so there is no need to always compute it at each call. The cache key is the buffer MD4 checksum, eventual collisions being handled. The cache location is the my_dist_data directory provided by File::HomeDir package. Default is a false value.

=item ANYTHINGELSE

Any other option that will be specific to the ECMAScript specific implementation. In practice these are:

=over

=item Template

Reference to a hash for template options. C.f. per-ECMAScript documentation.

=item StringNumericLiteral

Reference to a hash for StringNumericLiteral grammar options. C.f. per-ECMAScript documentation.

=item Pattern

Reference to a hash for Pattern grammar options. C.f. per-ECMAScript documentation.

=back

=back

=cut

# ----------------------------------------------------------------------------------------
sub new {
  my ($class, %opts) = @_;

  my $grammarName = $opts{grammarName} || 'ECMAScript-262-5';
  my $cache       = $opts{cache} // 0;

  my $self  = {
      _grammarName => $grammarName,
      _cache       => $cache
  };

  bless($self, $class);

  if (exists($opts{grammarName})) {
      delete($opts{grammarName});
  }
  if (exists($opts{cache})) {
      delete($opts{cache});
  }

  $self->_init(%opts);

  return $self;
}

sub _init {
    my ($self, %opts) = @_;

    $self->{_grammar} = MarpaX::Languages::ECMAScript::AST::Grammar->new($self->{_grammarName}, %opts);
}

# ----------------------------------------------------------------------------------------

=head2 describe($self)

Get a description of the G1 grammar. Returns a reference to hash, that has this structure: key => value, where

=over

=item key

'G1' or 'G0', with value being a reference to all rules, LHS being the first.

=back

=cut

sub describe {
    my ($self) = @_;

    my $impl = $self->{_grammar}->program->{impl};
    my %g0 = ();
    foreach ($impl->rule_ids('G0')) {
      $g0{$_} = [ map {$impl->symbol_name($_, 'G0')} $impl->rule_expand($_, 'G0') ];
    }
    my %g1 = ();
    foreach ($impl->rule_ids('G1')) {
      $g1{$_} = [ map {$impl->symbol_name($_, 'G1')} $impl->rule_expand($_, 'G1') ];
    }

    return { G0 => \%g0, G1 => \%g1 };
}

# ----------------------------------------------------------------------------------------

=head2 grammarAlias($self)

Returns the grammar alias, i.e. the one that is used within this distribution, corresponding to a true directory on the filesystem.

=cut

sub grammarAlias {
  my ($self) = @_;

  return $self->{_grammar}->grammarAlias;

}

# ----------------------------------------------------------------------------------------

=head2 templatePath($self)

Returns the templates absolute path.

=cut

sub templatePath {
  my ($self) = @_;

  return File::Spec->catdir(dirname(find_installed(__PACKAGE__)), 'AST', 'Grammar',  $self->{_grammar}->grammarAlias, 'Template');
}

# ----------------------------------------------------------------------------------------

=head2 parse($self, $source)

Get an AST from the ECMAScript source, pointed by $source. This method will call all the intermediary steps (lexical, transformation, evaluation) necessary to produce the AST. An AST is a reference to nested arrays, containing either:

=over

=item valuesAndRuleId

An entry describing a G1 rule (i.e. a non-terminal) is a reference to an array. The array contains the values of the right-hand side of the rule, followed by the rule Id.

=item startAndLengthAndValue

An entry describing a lexeme (i.e. a terminal) is a reference to an array. The array contains the start location in the stream, length and lexeme as perl's utf8 string.

=back

=cut

sub _getAndCheckHashFromCache {
  my ($self, $md4, $source, $astp, $fromCachep) = @_;

  my $rc = 0;

  my $fromCache = $CACHE->get($md4);
  if (defined($fromCache)) {
    my $clearCache = 1;
    my $store;
    if (ref($fromCache) eq 'HASH') {
      $store = $fromCache->{$source};
      if (defined($store)) {
        if (ref($store) eq 'HASH') {
          my $storeVersion = $store->{version};
          #
          # Trying to get from cache using the dev files will always clear the cache -;
          #
          if (defined($storeVersion) && defined($CURRENTVERSION)) {
            if (version::is_lax($storeVersion) && version::is_lax($CURRENTVERSION)) {
              if (version->parse($storeVersion) == version->parse($CURRENTVERSION)) {
                my $ast = $store->{ast};
                if (defined($ast)) {
                  $log->tracef('cache ok, storeVersion=%s', $storeVersion);
                  $rc = 1;
                  ${$astp} = $ast;
                  ${$fromCachep} = $fromCache;
                  $clearCache = 0;
                } else {
                  $log->tracef('cache ko, ast undefined');
                }
              } else {
                $log->tracef('cache ko, storeVersion %s != %s (current version)', $storeVersion, $CURRENTVERSION);
              }
            } else {
              #
              # In case versions are really garbled, use %s instead of %d
              #
              $log->tracef('cache ko, storeVersion %s (is_lax=%s), current version %s (is_lax=%s)', $storeVersion, version::is_lax($storeVersion) || '', $CURRENTVERSION, version::is_lax($CURRENTVERSION) || '');
            }
          } else {
            $log->tracef('cache ko, storeVersion %s, current version %s', $storeVersion || 'undefined', $CURRENTVERSION || 'undefined');
          }
        } else {
          $log->tracef('cache ko, store is a %s', ref($store));
        }
      } else {
        $log->tracef('cache ko, no entry for given source code');
      }
    } else {
      $log->tracef('cache ko, $fromCache is a %s', ref($fromCache));
    }
    if ($clearCache) {
      if (ref($fromCache) eq 'HASH') {
        #
        # Invalid data
        #
        if (defined($store)) {
          delete($fromCache->{$source});
          $CACHE->set($md4, $fromCache);
          $log->tracef('cache cleaned');
        }
      } else {
        #
        # Invalid cache
        #
        $CACHE->remove($md4);
        $log->tracef('cache removed');
      }
    }
  } else {
    $log->tracef('cache ko, no cache for md4 %s', $md4);
  }

  return $rc;
}

sub parse {
  my ($self, $source) = @_;

  my $parse = sub {
      my $grammar     = $self->{_grammar}->program->{grammar};
      my $impl        = $self->{_grammar}->program->{impl};

      return $grammar->parse($source, $impl)->value($impl);
  };

  #
  # If cache is enabled, compute the MD4 and check availability
  #
  my $ast;
  if ($self->{_cache}) {
    my $md4 = md4_hex($source);
    my $fromCache = {};
    if (! $self->_getAndCheckHashFromCache($md4, $source, \$ast, \$fromCache)) {
      $ast = &$parse();
      if (defined($CURRENTVERSION)) {
        $fromCache->{$source} = {ast => $ast, version => $CURRENTVERSION};
        $CACHE->set($md4, $fromCache);
      }
    }
  } else {
    $ast = &$parse();
  }

  return $ast;
}

# ----------------------------------------------------------------------------------------

=head2 template($self)

Return the generic template for this grammar. This template is doing nothing else but reproduce an ECMAScript source that, if parsed, would have an AST similar to the original source.

=cut

sub template {
  my ($self) = @_;

  return $self->{_grammar}->template;
}

# ----------------------------------------------------------------------------------------

=head2 stringNumericLiteral($self)

Return the generic stringNumericLiteral for this grammar.

=cut

# ----------------------------------------------------------------------------------------

sub stringNumericLiteral {
  my ($self) = @_;

  return $self->{_grammar}->stringNumericLiteral;
}

# ----------------------------------------------------------------------------------------

=head2 pattern($self)

Return the generic Pattern for this grammar.

=cut

sub pattern {
  my ($self) = @_;

  return $self->{_grammar}->pattern;
}

# ----------------------------------------------------------------------------------------

=head2 JSON($self)

Return the generic JSON for this grammar.

=cut

sub JSON {
  my ($self) = @_;

  return $self->{_grammar}->JSON;
}

# ----------------------------------------------------------------------------------------

=head2 URI($self)

Return the generic URI for this grammar.

=cut

sub URI {
  my ($self) = @_;

  return $self->{_grammar}->URI;
}

# ----------------------------------------------------------------------------------------

=head1 SEE ALSO

L<Log::Any>

L<Marpa::R2>

L<Digest::MD4>

L<CHI::Driver::File>

L<MarpaX::Languages::ECMAScript::AST::Grammar::ECMAScript_262_5>

=cut

1;
